### PR TITLE
Makes resillient functionsto Apps failures

### PR DESCRIPTION
**Pull request by vwraposo**
#### What is the purpose of this pull request?

Creates fallback strategy to Apps functions when it fails. 

**getAppsMetaInfos**:
Fallback: gets dependencies from disk cache

**getAppFile**:
1. Tries to get from disk
2. Makes call to Apps
3. Gets file from a previous version (same major)

#### Types of changes

* [ ] Bug fix (a non-breaking change which fixes an issue)
* [x] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: src/utils/apps/metaInfo.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+const hashMD5 = (text: string) =>
+  crypto
+    .createHash('md5')
+    .update(text)
+    .digest('hex')
+
+const locks: Record<string, AwaitLock> = {}
+
+const updateCache = async (account: string, workspace: string, dependencies: AppMetaInfo[], logger: Logger) => {
+  if (workspace !== 'fox2') {
```
Comment (Jeymisson) [inline src/utils/apps/metaInfo.ts]: 🔨

File: src/utils/apps/metaInfo.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  const key = getKey(account)
+  const hash = hashMD5(dependencies.toString())
+
+  try {
+    const storedDependencies = await diskCacheStorage.get(key) || ''
+      if (hash !== hashMD5(storedDependencies.toString())) {
+        await diskCacheStorage.set(key, dependencies)
+      }
+  } catch (error) {
+    logger.error({error, message: 'Apps disk cache update failed'})
```
Comment (Jeymisson) [inline src/utils/apps/metaInfo.ts]: Won't this generate an inconsistent state? Shouldn't we throw this error? 🤔

File: src/utils/apps/metaInfo.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  const lock = locks[account]
+  try {
+    const dependencies = await apps.getAppsMetaInfos()
+    await lock.acquireAsync()
+    updateCache(account, workspace, dependencies, logger)
+    return dependencies
+  } catch (error) {
+    if (workspace !== 'fox2') {
```
Comment (Jeymisson) [inline src/utils/apps/metaInfo.ts]: 🔨

File: src/utils/apps/appFile.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  if (fallbackVersion) {
+    const appFallbackVersion = `${appName}@${fallbackVersion}`
+    return apps.getAppFile(appFallbackVersion, path)
+      .then(prop('data'))
+      .then(toString)
+  }
+  return Promise.resolve('')
+}
+
+export const getAppFile = async (apps: Apps, app: string, path: string, cacheStorage: CacheLayer<string, string>, logger: Logger): Promise<string> => {
```
Comment (tlgimenes) [inline src/utils/apps/appFile.ts]: Why not making `Apps` cache storage public and retrieving it in here ?

File: src/utils/apps/appFile.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  const fallbackKey = getFallbackKey(appName, major)
+
+  const fallbackVersion = await cacheStorage.get(fallbackKey)
+  if (fallbackVersion) {
+    const appFallbackVersion = `${appName}@${fallbackVersion}`
+    return apps.getAppFile(appFallbackVersion, path)
+      .then(prop('data'))
+      .then(toString)
+  }
+  return Promise.resolve('')
```
Comment (tlgimenes) [inline src/utils/apps/appFile.ts]: I think you don't need the `Promise.resolve` in here

**Pull request merged**